Sources/FoundationInternationalization/Locale/Locale.swift

```diff
@@ -929,27 +929,138 @@ public struct Locale : Hashable, Equatable, Sendable {
+    private static let countryCodeKey = "kCFLocaleCountryCodeKey"
+    private static let variantCodeKey = "kCFLocaleVariantCodeKey"
+    private static let calendarKey = "kCFLocaleCalendarKey"
 
     /// Constructs an identifier from a dictionary of components.
     public static func identifier(fromComponents components: [String : String]) -> String {
-#if FOUNDATION_FRAMEWORK
-        // TODO: Port to Swift
-        // n.b. the CFLocaleCreateLocaleIdentifierFromComponents API is normally [String: String], but for 'convenience' allows a `Calendar` value for "kCFLocaleCalendarKey"/"calendar".
-        return CFLocaleCreateLocaleIdentifierFromComponents(kCFAllocatorSystemDefault, components as CFDictionary).rawValue as String
-#else
-        return ""
-#endif // FOUNDATION_FRAMEWORK
+        // Holds remaining keywords after we remove the CF-specific ones
+        var keywords = components
+        var result = ""
+        if let language = components[Self.languageCodeKey] {
+            result += language
+            keywords.removeValue(forKey: Self.languageCodeKey)
+        }
+        if let script = components[Self.scriptCodeKey] {
+            result += "_" + script
+            keywords.removeValue(forKey: Self.scriptCodeKey)
+        }
+        let country = components[Self.countryCodeKey]
+        let variant = components[Self.variantCodeKey]
+        
+        if country != nil || variant != nil {
+            result += "_"
+        }
+        
+        if let country {
+            result += country
+            keywords.removeValue(forKey: Self.countryCodeKey)
+        }
+        
+        if let variant {
+            result += "_" + variant
+            keywords.removeValue(forKey: Self.variantCodeKey)
+        }
+                
+        let corrected = Dictionary<String, String>(uniqueKeysWithValues: keywords.compactMap { key, value -> (String, String)? in
+            // Keys must be non-empty
+            guard !key.isEmpty else { return nil }
+            
+            // Identifier keywords must be ASCII a-z, A-Z, 0-9
+            // They are normalized to all-lowercase
+            var ok = true
+            var correctedKey = key.utf8.map { c in
+                if 0x41 <= c && c <= 0x5a {
+                    // A-Z
+                    // Convert to lowercase by adding 0x20
+                    return CChar(c + 0x20)
+                } else if 0x61 <= c && c <= 0x7a {
+                    // a-z
+                    return CChar(c)
+                } else if 0x30 <= c && c <= 0x39 {
+                    // 0-9
+                    return CChar(c)
+                } else {
+                    ok = false
+                    return CChar(0)
+                }
+            }
+            // null-terminate
+            correctedKey.append(CChar(0))
+            
+            guard ok else { return nil }
+            let correctedKeyString = String(cString: correctedKey)
```

Comment:
I believe calling `init(cString:)` involves another loop to find the buffer length. Since we're already iterating the utf8 buffer we could use another initializer that takes the count like this one, `String(unsafeUninitializedCapacity: count...)`? cc @glessard does that sound right to you?

Reply:
For what it's worth, the first time I wrote this I forgot the `null` termination and this initializer gave me an assert, which I appreciated. So it's a bit safer.